### PR TITLE
feat(azure): Add new checks related to Network service

### PR DESCRIPTION
**Pull request by pedrooot**
### Context

This pr adds 7 new checks related to network service.

### Description

The checks that are included in this pr are:

- [x] network_rdp_internet_access_restricted
- [x] network_ssh_internet_access_restricted
- [x] network_udp_internet_access_restricted
- [x] network_http_internet_access_restricted
- [x] network_flow_log_more_than_90_days
- [x] network_network_watcher_enabled
- [x] network_bastion_host_exists

### License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: prowler/providers/azure/services/network/network_service.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+                    subscription_locations = available_locations[subscription_id]
+                    security_groups[subscription].append(
+                        SecurityGroup(
+                            id=security_group.id,
+                            name=security_group.name,
+                            location=security_group.location,
+                            security_rules=security_group.security_rules,
+                            network_watchers=network_watchers,
+                            subscription_locations=subscription_locations,
+                            flow_logs=self.__get_flow_logs__(subscription),
```
Comment (MrCloudSec) [inline prowler/providers/azure/services/network/network_service.py]: can you do it up like the others?

File: prowler/providers/azure/services/network/network_bastion_host_exists/network_bastion_host_exists.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
+from prowler.lib.check.models import Check, Check_Report_Azure
+from prowler.providers.azure.services.network.network_client import network_client
+
+
+class network_bastion_host_exists(Check):
+    def execute(self) -> Check_Report_Azure:
+        findings = []
+        for subscription, bastion_hosts in network_client.bastion_hosts.items():
+            if bastion_hosts == []:
```
Comment (MrCloudSec) [inline prowler/providers/azure/services/network/network_bastion_host_exists/network_bastion_host_exists.py]:
```suggestion
            if not bastion_hosts:
```

File: prowler/providers/azure/services/network/network_bastion_host_exists/network_bastion_host_exists.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+                status = "FAIL"
+                status_extended = (
+                    f"Bastion Host from subscription {subscription} does not exist"
+                )
+            else:
+                status = "PASS"
+                status_extended = (
+                    f"Bastion Host from subscription {subscription} exists"
```
Comment (MrCloudSec) [inline prowler/providers/azure/services/network/network_bastion_host_exists/network_bastion_host_exists.py]: Please, add here the bastion hosts that are available

File: prowler/providers/azure/services/network/network_flow_log_more_than_90_days/network_flow_log_more_than_90_days.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+            for security_group in security_groups:
+                report = Check_Report_Azure(self.metadata())
+                report.subscription = subscription
+                report.resource_name = security_group.name
+                report.resource_id = security_group.id
+                if security_group.flow_logs and security_group.flow_logs.__len__() > 0:
+                    report.status = "PASS"
+                    report.status_extended = f"Security Group {security_group.name} from subscription {subscription} has flow logs enabled for more than 90 days"
+                    has_failed = False
+                    if not has_failed:
```
Comment (MrCloudSec) [inline prowler/providers/azure/services/network/network_flow_log_more_than_90_days/network_flow_log_more_than_90_days.py]:
```suggestion
```

File: prowler/providers/azure/services/network/network_flow_log_more_than_90_days/network_flow_log_more_than_90_days.py [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+class network_flow_log_more_than_90_days(Check):
+    def execute(self) -> Check_Report_Azure:
+        findings = []
+        for subscription, security_groups in network_client.security_groups.items():
+            for security_group in security_groups:
+                report = Check_Report_Azure(self.metadata())
+                report.subscription = subscription
+                report.resource_name = security_group.name
+                report.resource_id = security_group.id
+                if security_group.flow_logs and security_group.flow_logs.__len__() > 0:
```
Comment (MrCloudSec) [inline prowler/providers/azure/services/network/network_flow_log_more_than_90_days/network_flow_log_more_than_90_days.py]:
```suggestion
                if security_group.flow_logs:
```

File: prowler/providers/azure/services/network/network_http_internet_access_restricted/network_http_internet_access_restricted.py [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+        for subscription, security_groups in network_client.security_groups.items():
+            for security_group in security_groups:
+                report = Check_Report_Azure(self.metadata())
+                report.subscription = subscription
+                report.resource_name = security_group.name
+                report.resource_id = security_group.id
+                report.status = "PASS"
+                report.status_extended = f"Security Group {security_group.name} from subscription {subscription} has HTTP internet access restricted."
+                rule_fail_condition = any(
+                    rule.destination_port_range == "80"
```
Comment (MrCloudSec) [inline prowler/providers/azure/services/network/network_http_internet_access_restricted/network_http_internet_access_restricted.py]: Please, review if the rule accepts all ports or a range of ports that includes port 80.

File: prowler/providers/azure/services/network/network_network_watcher_enabled/network_network_watcher_enabled.metadata.json [hunk 7]
```diff
@@ -0,0 +1,30 @@
+{
+  "Provider": "azure",
+  "CheckID": "network_network_watcher_enabled",
```
Comment (MrCloudSec) [inline prowler/providers/azure/services/network/network_network_watcher_enabled/network_network_watcher_enabled.metadata.json]:
```suggestion
  "CheckID": "network_watcher_enabled",
```

File: prowler/providers/azure/services/network/network_service.py [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+########################## SQLServer
+class Network(AzureService):
+    def __init__(self, audit_info):
+        super().__init__(NetworkManagementClient, audit_info)
+        self.token = self.__get_token__(audit_info)
+        self.security_groups = self.__get_security_groups__(self.token)
+        self.bastion_hosts = self.__get_bastion_hosts__()
+
+    def __get_security_groups__(self, token):
+        logger.info("SQL Server - Getting Network Security Groups...")
```
Comment (MrCloudSec) [inline prowler/providers/azure/services/network/network_service.py]: Please, remove SQL Server from the INFO logs.

File: prowler/providers/azure/services/network/network_service.py [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+        security_groups = {}
+        for subscription, client in self.clients.items():
+            try:
+                security_groups.update({subscription: []})
+                security_groups_list = client.network_security_groups.list_all()
+                available_locations = {}
+                network_watchers = self.__get_network_watchers__(client, subscription)
```
Comment (MrCloudSec) [inline prowler/providers/azure/services/network/network_service.py]: Please, do a separate function and class for Network Watchers.

File: prowler/providers/azure/services/network/network_service.py (hunk 1 above)
Comment (MrCloudSec) [inline prowler/providers/azure/services/network/network_service.py]: add it to the new Network Watchers class

File: prowler/providers/azure/services/network/network_rdp_internet_access_restricted/network_rdp_internet_access_restricted.py [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+        for subscription, security_groups in network_client.security_groups.items():
+            for security_group in security_groups:
+                report = Check_Report_Azure(self.metadata())
+                report.subscription = subscription
+                report.resource_name = security_group.name
+                report.resource_id = security_group.id
+                report.status = "PASS"
+                report.status_extended = f"Security Group {security_group.name} from subscription {subscription} has RDP internet access restricted."
+                rule_fail_condition = any(
+                    rule.destination_port_range == "3389"
```
Comment (MrCloudSec) [inline prowler/providers/azure/services/network/network_rdp_internet_access_restricted/network_rdp_internet_access_restricted.py]: Same as HTTP check

File: prowler/providers/azure/services/network/network_ssh_internet_access_restricted/network_ssh_internet_access_restricted.py [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+        for subscription, security_groups in network_client.security_groups.items():
+            for security_group in security_groups:
+                report = Check_Report_Azure(self.metadata())
+                report.subscription = subscription
+                report.resource_name = security_group.name
+                report.resource_id = security_group.id
+                report.status = "PASS"
+                report.status_extended = f"Security Group {security_group.name} from subscription {subscription} has SSH internet access restricted."
+                rule_fail_condition = any(
+                    rule.destination_port_range == "22"
```
Comment (MrCloudSec) [inline prowler/providers/azure/services/network/network_ssh_internet_access_restricted/network_ssh_internet_access_restricted.py]: Same as previous checks

File: pyproject.toml [hunk 12]
```diff
@@ -53,6 +53,7 @@ schema = "0.7.5"
 shodan = "1.31.0"
 slack-sdk = "3.26.2"
 tabulate = "0.9.0"
+azure-mgmt-network = "^25.2.0"
```
Comment (MrCloudSec) [inline pyproject.toml]:
```suggestion
azure-mgmt-network = "25.2.0"
```

**Review by MrCloudSec — CHANGES_REQUESTED** [review]
Please, review my comments.

File: prowler/providers/azure/services/network/network_bastion_host_exists/network_bastion_host_exists.py (hunk 2 above)
Comment (pedrooot) [reply prowler/providers/azure/services/network/network_bastion_host_exists/network_bastion_host_exists.py]: Done

File: prowler/providers/azure/services/network/network_service.py (hunk 1 above)
Comment (pedrooot) [reply prowler/providers/azure/services/network/network_service.py]: Done!

File: prowler/providers/azure/services/network/network_bastion_host_exists/network_bastion_host_exists.py (hunk 3 above)
Comment (pedrooot) [reply prowler/providers/azure/services/network/network_bastion_host_exists/network_bastion_host_exists.py]: Done!

File: prowler/providers/azure/services/network/network_flow_log_more_than_90_days/network_flow_log_more_than_90_days.py (hunk 4 above)
Comment (pedrooot) [reply prowler/providers/azure/services/network/network_flow_log_more_than_90_days/network_flow_log_more_than_90_days.py]: Done

File: prowler/providers/azure/services/network/network_flow_log_more_than_90_days/network_flow_log_more_than_90_days.py (hunk 5 above)
Comment (pedrooot) [reply prowler/providers/azure/services/network/network_flow_log_more_than_90_days/network_flow_log_more_than_90_days.py]: Done

File: prowler/providers/azure/services/network/network_http_internet_access_restricted/network_http_internet_access_restricted.py (hunk 6 above)
Comment (pedrooot) [reply prowler/providers/azure/services/network/network_http_internet_access_restricted/network_http_internet_access_restricted.py]: Done

File: prowler/providers/azure/services/network/network_network_watcher_enabled/network_network_watcher_enabled.metadata.json (hunk 7 above)
Comment (pedrooot) [reply prowler/providers/azure/services/network/network_network_watcher_enabled/network_network_watcher_enabled.metadata.json]: Done

File: prowler/providers/azure/services/network/network_service.py (hunk 8 above)
Comment (pedrooot) [reply prowler/providers/azure/services/network/network_service.py]: Done

File: prowler/providers/azure/services/network/network_service.py (hunk 9 above)
Comment (pedrooot) [reply prowler/providers/azure/services/network/network_service.py]: Done

File: prowler/providers/azure/services/network/network_service.py (hunk 1 above)
Comment (pedrooot) [reply prowler/providers/azure/services/network/network_service.py]: Done

File: prowler/providers/azure/services/network/network_rdp_internet_access_restricted/network_rdp_internet_access_restricted.py (hunk 10 above)
Comment (pedrooot) [reply prowler/providers/azure/services/network/network_rdp_internet_access_restricted/network_rdp_internet_access_restricted.py]: Done

File: prowler/providers/azure/services/network/network_ssh_internet_access_restricted/network_ssh_internet_access_restricted.py (hunk 11 above)
Comment (pedrooot) [reply prowler/providers/azure/services/network/network_ssh_internet_access_restricted/network_ssh_internet_access_restricted.py]: Done

File: pyproject.toml (hunk 12 above)
Comment (pedrooot) [reply pyproject.toml]: Done

File: prowler/providers/azure/services/network/network_service.py [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+                for security_group in security_groups_list:
+                    subscription_id = security_group.id.split("/")[2]
+                    subscription_locations = audit_info.locations[subscription_id]
+                    security_groups[subscription].append(
+                        SecurityGroup(
+                            id=security_group.id,
+                            name=security_group.name,
+                            location=security_group.location,
+                            security_rules=security_group.security_rules,
+                            subscription_locations=subscription_locations,
```
Comment (MrCloudSec) [inline prowler/providers/azure/services/network/network_service.py]:
```suggestion
```

File: prowler/providers/azure/services/network/network_service.py [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+    def __get_security_groups__(self, audit_info):
+        logger.info("Network - Getting Network Security Groups...")
+        security_groups = {}
+        for subscription, client in self.clients.items():
+            try:
+                security_groups.update({subscription: []})
+                security_groups_list = client.network_security_groups.list_all()
+                for security_group in security_groups_list:
+                    subscription_id = security_group.id.split("/")[2]
+                    subscription_locations = audit_info.locations[subscription_id]
```
Comment (MrCloudSec) [inline prowler/providers/azure/services/network/network_service.py]:
```suggestion
```

File: prowler/providers/azure/services/network/network_service.py [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+    flow_logs: list
+
+
+@dataclass
+class SecurityGroup:
+    id: str
+    name: str
+    location: str
+    security_rules: list
+    subscription_locations: list
```
Comment (MrCloudSec) [inline prowler/providers/azure/services/network/network_service.py]:
```suggestion
```

File: prowler/providers/azure/services/network/network_watcher_enabled/network_watcher_enabled.py [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+
+class network_watcher_enabled(Check):
+    def execute(self) -> Check_Report_Azure:
+        findings = []
+        nw_locations = []
+        for subscription, network_watchers in network_client.network_watchers.items():
+            for network_watcher in network_watchers:
+                nw_locations.append(network_watcher.location)
+        for subscription, security_groups in network_client.security_groups.items():
+            for security_group in security_groups:
```
Comment (MrCloudSec) [inline prowler/providers/azure/services/network/network_watcher_enabled/network_watcher_enabled.py]: Iterate with the locations that are in audit info, please

File: tests/providers/azure/services/network/network_service_test.py [hunk 17]
```diff
@@ -0,0 +1,143 @@
...
+            network.security_groups[AZURE_SUBSCRIPTION][0].__class__.__name__
+            == "SecurityGroup"
+        )
+        assert network.security_groups[AZURE_SUBSCRIPTION][0].id == "id"
+        assert network.security_groups[AZURE_SUBSCRIPTION][0].name == "name"
+        assert network.security_groups[AZURE_SUBSCRIPTION][0].location == "location"
+        assert network.security_groups[AZURE_SUBSCRIPTION][0].security_rules == []
+        assert network.security_groups[AZURE_SUBSCRIPTION][
+            0
+        ].subscription_locations == ["location"]
```
Comment (MrCloudSec) [inline tests/providers/azure/services/network/network_service_test.py]:
```suggestion
```

File: prowler/providers/azure/azure_provider.py [hunk 18]
```diff
@@ -212,3 +213,31 @@ def get_identity(self):
 
     def get_region_config(self):
         return self.region_config
+
+    def get_locations(self, credentials, region_config):
```
Comment (MrCloudSec) [inline prowler/providers/azure/azure_provider.py:217]: Please, add tests to this function

File: prowler/providers/azure/services/network/network_service.py (hunk 13 above)
Comment (pedrooot) [reply prowler/providers/azure/services/network/network_service.py]: Done

File: prowler/providers/azure/services/network/network_service.py (hunk 14 above)
Comment (pedrooot) [reply prowler/providers/azure/services/network/network_service.py]: Done

File: prowler/providers/azure/services/network/network_service.py (hunk 15 above)
Comment (pedrooot) [reply prowler/providers/azure/services/network/network_service.py]: Done

File: prowler/providers/azure/services/network/network_watcher_enabled/network_watcher_enabled.py (hunk 16 above)
Comment (pedrooot) [reply prowler/providers/azure/services/network/network_watcher_enabled/network_watcher_enabled.py]: Done

File: tests/providers/azure/services/network/network_service_test.py (hunk 17 above)
Comment (pedrooot) [reply tests/providers/azure/services/network/network_service_test.py]: Done

File: prowler/providers/azure/azure_provider.py (hunk 18 above)
Comment (pedrooot) [reply prowler/providers/azure/azure_provider.py:217]: Done

**Review by MrCloudSec — DISMISSED** [review]
Awesome PR!!

**Review by MrCloudSec — DISMISSED** [review]

**Review by MrCloudSec — APPROVED** [review]

**Pull request merged**